File: lib/middleware/user_session.py
from lib.settings import Settings
from lib.db.models.session import Session
from lib.db.models.user import User


async def user_session(request):
    request.ctx.user = None
    request.ctx.session = None

    session_id = request.cookies.get(Settings.session_cookie_name, None)
    if session_id is None:
        return
    sess = await Session.find(session_id)
    if sess is None:
        return

    request.ctx.session = sess
    request.ctx.user = await User.find(sess.user_id)

File: MoonlightDashboard/MoonlightDashboard/Middleware/UserSession.cs
using MoonlightDashboard.Lib;

namespace MoonlightDashboard.Middleware
{

    public class UserSession
    {
        public const string cookieName = "sessionid";

        private readonly RequestDelegate _next;

        public UserSession(RequestDelegate next)
        {
            _next = next;
        }

        public async Task InvokeAsync(HttpContext context)
        {
            context.SetCurrentSession(null);
            context.SetCurrentUser(null);

            using var db = context.RequestServices.GetRequiredService<Database.DatabaseContext>();
            var sentSessionId = context.Request.Cookies[cookieName];
            if(!string.IsNullOrEmpty(sentSessionId))
            {
                var session = db.Sessions.FirstOrDefault(s => s.Id == sentSessionId);
                if(session != null)
                {
                    var user = db.Users.FirstOrDefault(u => u.Id == session.UserId);
                    if (user != null)
                    {
                        context.SetCurrentSession(session);
                        context.SetCurrentUser(user);
                    }
                }
            }
            
            await _next(context);
        }
    }

    public static class UserSessionExtensions
    {
        public static IApplicationBuilder UseUserSessions(
            this IApplicationBuilder builder)
        {
            return builder.UseMiddleware<UserSession>();
        }
    }

}
